archive/miu_language/decision_suf.lean

```diff
@@ -0,0 +1,382 @@
+`derivable en`. Decidability of `derivable en` is an easy consequence.
+
+The proof proceeds by induction on the `count U` of `en`.
+
+We tackle first the base case of the induction. This requires auxiliary results giving
+conditions under which  `count I ys = length ys`.
+-/
+
+
+/--
+If an `miustr` has a zero `count U` and contains no `M`, then its `count I` is its length.
+-/
+lemma count_I_eq_length_of_count_U_zero_and_neg_mem {ys : miustr} (hu : count U ys = 0)
+  (hm : M ∉ ys) : count I ys = length ys :=
+begin
+  induction ys with x xs hxs,
+  { refl, },
+  cases x,
+  { exfalso, exact hm (mem_cons_self M xs), }, -- case `x = M` gives a contradiction.
+  { rw [count_cons, if_pos (rfl),length,succ_eq_add_one,succ_inj'], -- case `x = I`
+    apply hxs,
+    { simpa only [count], },
+    simp only [mem_cons_iff,false_or] at hm, exact hm, },
+  exfalso, simp only [count,countp_cons_of_pos] at hu,  -- case `x = U` gives a contradiction.
+  exact succ_ne_zero _ hu,
+end
+
+/--
+`base_case_suf` is the base case of the sufficiency result.
+-/
+lemma base_case_suf (en : miustr) (h : decstr en) (hu : count U en = 0) : derivable en :=
+begin
+  rcases h with ⟨⟨mhead, nmtail⟩, hi ⟩,
+  have : en ≠ nil,
+  { intro k, simp [k,count] at hi, contradiction, },
+  rcases (exists_cons_of_ne_nil this) with ⟨y,ys,rfl⟩,
+  rw head at mhead,
+  rw mhead at *,
+  suffices  : ∃ c, repeat I c = ys ∧ (c % 3 = 1 ∨ c % 3 = 2),
+  { rcases this with ⟨c, hysr, hc⟩,
+    rw ←hysr,
+    exact der_repeat_I_of_mod3 c hc, },
+  simp only [count] at *,
+  use (count I ys),
+  split,
+  { apply repeat_count_eq_of_count_eq_length,
+    apply count_I_eq_length_of_count_U_zero_and_neg_mem,
+    { exact hu, },
+    exact nmtail, },
+  exact hi,
+end
+
+
+
+
+/-!
+Before continuing to the proof of the induction step, we need other auxiliary results that
+relate to `count U`.
+-/
+lemma mem_of_count_U_eq_succ {xs : miustr} {k : ℕ} (h : count U xs = succ k) : U ∈ xs :=
+begin
+  induction xs with z zs hzs,
+  { exfalso, rw count at h, contradiction, },
+  simp only [mem_cons_iff],
+  cases z,
+    repeat -- cases `z = M` and `z=I`
+    { right, apply hzs, simp only [count,countp,if_false] at h, rw ←h, refl, },
+  left, refl, -- case `z = U`
+end
+
+
+
+lemma eq_append_cons_U_of_count_U_pos {k : ℕ} {zs : miustr} (h : count U zs = succ k) :
+∃ (as bs : miustr), (zs = as ++ U :: bs) :=
+begin
+  apply mem_split,
+  exact mem_of_count_U_eq_succ h,
+end
+
+
+/--
+`ind_hyp_suf` is the inductive step of the sufficiency result.
+ -/
+lemma ind_hyp_suf (k : ℕ) (ys : miustr) (hu : count U ys = succ k) (hdec : decstr ys) :
+∃ (as bs : miustr), (ys = M::as ++ U:: bs) ∧ (count U (M::as ++ [I,I,I] ++ bs) = k) ∧
+  decstr (M::as ++ [I,I,I] ++ bs) :=
+begin
+  rcases hdec with ⟨⟨mhead,nmtail⟩, hic⟩,
+  have : ys ≠ nil,
+  { intro k, simp [k,count] at hic, contradiction, },
+  rcases (exists_cons_of_ne_nil this) with ⟨z,zs,rfl⟩,
+  rw head at mhead,
+  rw mhead at *,
+  simp [count] at *,
```

Comment:
Please try to avoid nonterminal calls to `simp`, here and below. `simp only` is fine.